### DISCUSSION TITLE
Inconsistent  type inference for date add/subtract interval operations

### DISCUSSION THREAD

**Discussion by LeonidChistov** [substrait-io/substrait#574] [General]
Following operations are defined in Substrait prototypes:

add(date, interval_year): -> timestamp
add(date, interval_day): -> timestamp
subtract(date, interval_year): -> date
subtract(date, interval_day): -> date

Is there a reason why `add` and `subtract` should behave differently wrt to result type?

**Comment by EpsilonPrime** [comment]
I don't believe there is a reason.  The add functions taking a date should return a date (the same type as the input).

**Reply by LeonidChistov** [reply]
IMHO, following type inference would make more sense:

add(date, interval_year): -> date
add(date, interval_day): -> timestamp
subtract(date, interval_year): -> date
subtract(date, interval_day): -> timestamp

Because otherwise there is a high chance that we just loose "interval_day" value except for the "day" part.

But from the other point of view, someone who does not want to loose precision, may cast date to timestamp before adding/subtracting interval.

Do you think it make sense to create Issue based on that discussion?

**Comment by EpsilonPrime** [comment]
Your analysis makes sense, it should be timestamp.  I did find an old PR (#337) that corrected subtract's type, presumably they missed add at that time.

We could go with an issue or straight into a PR.

**Reply by LeonidChistov** [reply]
https://github.com/substrait-io/substrait/issues/576
